### PR TITLE
expressions highlighter

### PR DESCRIPTION
**Pull request by sylviameows**
implements a feature previously provided by recode (the "expression highlighter").

**what it does**:
edits minecraft's chat box field to provide syntax highlighting for minimessage and DF expressions in commands like `/plot name`, `/txt`, and `/lore` as well as when editing variable items like strings, styled text, numbers, and variables.

examples:
example of minimessage highlighting
example of expression highlighting

also includes a config which allows you to customize how the expression highlighter appears:
example of config

there may be some strange quirks with it though that may need more ironing out, but in its current state it is definitely passable

### CODE REVIEW

File: src/main/java/dev/dfonline/codeclient/config/Config.java [hunk 1]
```diff
@@ -81,6 +81,11 @@ public class Config {
     public CPUDisplayCorner CPUDisplayCornerOption = CPUDisplayCorner.TOP_LEFT;
     public boolean HideScopeChangeMessages = true;
 
+    public boolean HighlighterEnabled = true; // whether expression highlighting is enabled.
+    public boolean HighlightExpressions = true; // if the highlighter marks up df expressions
+    public boolean HighlightMiniMessage = true; // if the highlighter marks up minimessage.
+    public int MiniMessageTagColor = 0x808080; // the color of minimessage tags in the edit box.
```
Comment (GeorgeRNG) [inline src/main/java/dev/dfonline/codeclient/config/Config.java:87]: nothing else has comments. for consistency remove them or we could move to adding javadoc to each one, although they do have their description under getLibConfig

File: src/main/java/dev/dfonline/codeclient/dev/highlighter/ExpressionHighlighter.java [hunk 2]
```diff
@@ -0,0 +1,350 @@
...
+            TextColor.fromRgb(0x5e77f7),
+            TextColor.fromRgb(0xca64fa),
+            TextColor.fromRgb(0xff4242)
+    );
+
+    private final TextColor ERROR_COLOR = TextColor.fromFormatting(Formatting.RED);
+
+    @Override
+    public boolean enabled() {
+        return CodeClient.location instanceof Dev && Config.getConfig().HighlighterEnabled;
```
Comment (GeorgeRNG) [inline src/main/java/dev/dfonline/codeclient/dev/highlighter/ExpressionHighlighter.java:68]: I believe this function is for if it's enabled by the user; not if the state means it's active: https://github.com/DFOnline/CodeClient/blob/d7b56fdd1d56a3a97f140a04516b3b545dca210a/src/main/java/dev/dfonline/codeclient/Feature.java#L21-L27

**Review by GeorgeRNG — CHANGES_REQUESTED** [review]

File: src/main/java/dev/dfonline/codeclient/dev/highlighter/ExpressionHighlighter.java (hunk 2 above)
Comment (sylviameows) [reply src/main/java/dev/dfonline/codeclient/dev/highlighter/ExpressionHighlighter.java]: i believe i have fixed this, for some reason i thought this was how the other features used this method

**Comment by RedVortexDev** [comment]
image
Seems like decoration tags apply to the tags after it

**Comment by sylviameows** [comment]
> image 
> Seems like decoration tags apply to the tags after it

I tried to fix this, but escaping decoration tags is not as straightforward as one might assume, if anyone else would like to take a crack at figuring it out, i kept my attempt inside the code but commented out. (this also isnt detrimental to user experience imo)

> image 
> characters after wrong % codes start to leak the color code

This bug has been fixed.

**Review by GeorgeRNG — APPROVED** [review]

**Pull request merged**